packages/@n8n/nodes-langchain/nodes/tools/ToolHttpRequest/utils.ts

```diff
@@ -0,0 +1,793 @@
+};
+
+export const extractParametersFromText = (
+	placeholders: PlaceholderDefinition[],
+	text: string,
+	sendIn: SendIn,
+	key?: string,
+): ToolParameter[] => {
+	if (typeof text !== 'string') return [];
+
+	const parameters = extractPlaceholders(text);
+
+	if (parameters.length) {
+		// eslint-disable-next-line @typescript-eslint/no-use-before-define
+		const inputParameters = prepareParameters(
+			parameters.map((name) => ({
+				name,
+				valueProvider: 'modelRequired',
+			})),
+			placeholders,
+			'keypair',
+			sendIn,
+			'',
+		);
+
+		return key
+			? inputParameters.parameters.map((p) => ({ ...p, key }))
+			: inputParameters.parameters;
+	}
+
+	return [];
+};
+
+function prepareParameters(
+	rawParameters: RawParametersValues,
+	placeholders: PlaceholderDefinition[],
+	parametersInputType: 'model' | 'keypair' | 'json',
+	sendIn: SendIn,
+	modelInputDescription: string,
+	jsonWithPlaceholders?: string,
+): { parameters: ToolParameter[]; values: IDataObject } {
+	const parameters: ToolParameter[] = [];
+	const values: IDataObject = {};
+
+	if (parametersInputType === 'model') {
+		return {
+			parameters: [
+				{
+					name: sendIn,
+					required: true,
+					type: 'json',
+					description: modelInputDescription,
+					sendIn,
+				},
+			],
+			values: {},
+		};
+	}
+
+	if (parametersInputType === 'keypair') {
+		for (const entry of rawParameters) {
+			if (entry.valueProvider.includes('model')) {
+				const placeholder = placeholders.find((p) => p.name === entry.name);
+
+				const parameter: ToolParameter = {
+					name: entry.name,
+					required: entry.valueProvider === 'modelRequired',
+					sendIn,
+				};
+
+				if (placeholder) {
+					parameter.type = placeholder.type;
+					parameter.description = placeholder.description;
+				}
+
+				parameters.push(parameter);
+			} else {
+				// if value has placeholders push them to parameters
+				parameters.push(
+					...extractParametersFromText(placeholders, entry.value as string, sendIn, entry.name),
+				);
+				values[entry.name] = entry.value; //push to user provided values
+			}
+		}
+	}
+
+	if (parametersInputType === 'json' && jsonWithPlaceholders) {
+		parameters.push(
+			...extractParametersFromText(placeholders, jsonWithPlaceholders, sendIn, `${sendIn + 'Raw'}`),
+		);
+	}
+
+	return {
+		parameters,
+		values,
+	};
+}
+
+const MODEL_INPUT_DESCRIPTION = {
+	qs: 'Query parameters for request as key value pairs',
+	headers: 'Headers parameters for request as key value pairs',
+	body: 'Body parameters for request as key value pairs',
+};
+
+export const updateParametersAndOptions = (
+	ctx: IExecuteFunctions,
+	itemIndex: number,
+	toolParameters: ToolParameter[],
+	placeholdersDefinitions: PlaceholderDefinition[],
+	requestOptions: IHttpRequestOptions,
+	rawRequestOptions: { [key: string]: string },
+	requestOptionsProperty: 'headers' | 'qs' | 'body',
+	inputTypePropertyName: string,
+	jsonPropertyName: string,
+	parametersPropertyName: string,
+) => {
```

Comment:
IMO if function has more than 2 arguments, it should accept an object instead. Because it's very hard to understand what is what just by looking at the invocation:
![CleanShot 2024-06-17 at 11 23 42@2x](https://github.com/n8n-io/n8n/assets/12657221/87069cd6-c269-41a5-b331-1adc7e4130f4)